pkg/reconciler/namespace/controller.go

```diff
@@ -0,0 +1,256 @@
+	}
+	klog.Infof("Starting workers")
+	<-c.stopCh
+	klog.Infof("Stopping workers")
+}
+
+func (c *Controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	var err error
+	if strings.HasPrefix(key, "namespace::") {
+		err = c.processNamespace(ctx, strings.TrimPrefix(key, "namespace::"))
+	} else if strings.HasPrefix(key, "cluster::") {
+		err = c.processCluster(ctx, strings.TrimPrefix(key, "cluster::"))
+	} else {
+		err = c.process(ctx, key)
+	}
+	c.handleErr(err, key)
+	return true
+}
+
+func (c *Controller) handleErr(err error, key string) {
+	// Reconcile worked, nothing else to do for this workqueue item.
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	// Re-enqueue up to 5 times.
+	num := c.queue.NumRequeues(key)
+	if num < 5 {
+		klog.Errorf("Error reconciling key %q, retrying... (#%d): %v", key, num, err)
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	// Give up and report error elsewhere.
+	c.queue.Forget(key)
+	runtime.HandleError(err)
+	klog.Infof("Dropping key %q after failed retries: %v", key, err)
+}
+
+// key is gvr::namespace/name
+func (c *Controller) process(ctx context.Context, key string) error {
+	parts := strings.SplitN(key, "::", 2)
+	if len(parts) != 2 {
+		return fmt.Errorf("invalid key: %s", key)
+	}
+	gvrstr := parts[0]
+	gvr, _ := schema.ParseResourceArg(gvrstr)
+	if gvr == nil {
+		return fmt.Errorf("Error parsing GVR %q", gvrstr)
+	}
+	key = parts[1]
+
+	obj, exists, err := c.indexer.GetByKey(key)
+	if err != nil {
+		return err
+	}
+
+	if !exists {
+		klog.Infof("Object with key %q was deleted", key)
+		return nil
+	}
+	current := obj.(*unstructured.Unstructured)
+	previous := current.DeepCopy()
+
+	if err := c.reconcileResource(ctx, current, gvr); err != nil {
+		return err
+	}
+
+	// If the object being reconciled changed as a result, update it.
+	if !equality.Semantic.DeepEqual(previous, current) {
+		_, uerr := c.dynClient.Resource(*gvr).Namespace(current.GetNamespace()).Update(ctx, current, metav1.UpdateOptions{})
+		return uerr
+	}
+	return nil
+}
+
+func (c *Controller) processNamespace(ctx context.Context, key string) error {
+	obj, exists, err := c.indexer.GetByKey(key)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		klog.Infof("Object with key %q was deleted", key)
+	}
+	current := obj.(*corev1.Namespace)
+	previous := current.DeepCopy()
+
+	if err := c.reconcileNamespace(ctx, current); err != nil {
+		return err
+	}
+
+	// If the namespace changed as a result, update it.
+	if !equality.Semantic.DeepEqual(previous, current) {
+		_, uerr := c.kubeClient.CoreV1().Namespaces().Update(ctx, current, metav1.UpdateOptions{})
```

Comment:
I'd recommend Patch over Update to avoid accidental field pruning (this client vendors the go struct for Namespace @ v1.22.x, server upgrades to v1.23+ which adds a new field to Namespace -> this client doing an update will result in that newer field being dropped).

Reply:
Would you agree @smarterclayton ?

Reply:
Good call, and done in the latest revision. It turned out to be quite a bit simpler too since the patch bytes to describe setting the label is the same whether it's a resource or a namespace.

I'm now actually doing the Patch inside `reconcileResource`, instead of doing it implicitly outside, since updating the label is the only expected modification of the object, and this now matches `reconcileNamespace` and `reconcileCluster` semantics.

The only downside is that testing this is a bit painful since I have to get a fake clientset and a fake dynamic client to agree about what fake objects I've put into both of them, and I can't seem to figure out how to do that. I'll keep trying though.